OpenZFS 7252 - create long versions of ZFS send / receive options

bunder2015 opened the pull request:
Authored by: Dan Kimmel <dan.kimmel@delphix.com>
Reviewed by: George Wilson <george.wilson@delphix.com>
Reviewed by: John Kennedy <john.kennedy@delphix.com>
Reviewed by: Matthew Ahrens <mahrens@delphix.com>
Reviewed by: Paul Dagnelie <pcd@delphix.com>
Reviewed by: Pavel Zakharov <pavel.zakharov@delphix.com>
Reviewed by: Sebastien Roy <sebastien.roy@delphix.com>
Reviewed by: David Quigley <dpquigl@davequigley.com>
Reviewed by: Thomas Caputi <tcaputi@datto.com>
Approved by: Dan McDonald <danmcd@omniti.com>
Ported by: bunder2015 <omfgbunder@gmail.com>

Porting over some remaining changes from OpenZFS commit 5602294

OpenZFS-issue: https://www.illumos.org/issues/7628
OpenZFS-commit: https://github.com/openzfs/openzfs/commit/5602294

### Motivation and Context
Thought this would be an easy first attempt at porting, found it on the tracking page.

### How Has This Been Tested?
Please be kind to me buildbots and reviewers. 😃 
edit: I'll try to get a personal build box VM going this week, sorry for the buildbot noise.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

bunder2015 commented:
Okay I think I got it right this time...  but something seems off with that new formatting checker...

`error: missing "Ported-by"
error: missing "Ported-by"`

It's there... but what is also odd, is it's saying that twice.

dinatale2 commented:
@bunder2015 I noticed that. I'm looking into it. Thanks for the heads up.

dinatale2 commented:
@bunder2015 The issues you were seeing should be corrected once #5952 lands. Sorry for the trouble!

bunder2015 commented:
Looks like I missed some illumos-isms in the test scripts.  I'll try to get these cleaned up.

bunder2015 commented on tests/zfs-tests/tests/functional/rsend/rsend.kshlib:
```diff
@@ -551,3 +551,120 @@ function test_fs_setup
+		known=${feature[$flag]}
+		[[ -z $known ]] && log_fail "Unknown feature: $flag"
+
+		derived=$($ECHO "$flags & ${feature[$flag]} = X" | mdb | $SED 's/ //g')
```
I need a replacement for mdb here, I tried using ksh directly but it doesn't provide the correct output

bunder2015 commented on tests/zfs-tests/tests/functional/rsend/send-cD.ksh:
```diff
@@ -0,0 +1,77 @@
+log_must $ZFS create -o compress=lz4 $recvfs
+typeset dir=$(get_prop mountpoint $sendfs)
+# Don't use write_compressible: we want compressible but undedupable data here.
+log_must $CP /kernel/genunix $dir/file
```
Any recommendations for a replacement file to use that's available on Linux?

bunder2015 replied:
I've been trying to think of a file that would be on every buildbot and that has a reasonable size, would something like awk, grep, mv, ls, tar, ps be large enough for this?  Most of these are over 100kb in size.  If we need bigger, ksh and git are around 2mb in size.

dinatale2 replied:
`mkfile` may be useful to make a file in the size you want.

bunder2015 replied:
Do you know if mkfile puts any content in the file, or if its just an empty file?  I imagine that would have the same problem as `write_compressible`.

dinatale2 replied:
Hmm, it may make a file full of 0s. Another option may be to use `dd` and with your input stream being `/dev/urandom` or something similar.

bunder2015 replied:
Unfortunately, urandom doesn't compress.

`Compressed 4194304 bytes into 4194323 bytes ==> 100.00%`

behlendorf replied:
You can use `write_file` which will write a single non-zero compressible pattern.

tonyhutter replied:
Fun fact, if you pass `R` to `file_write`'s `-d` option, it will write psudorandom data:

```./tests/zfs-tests/cmd/file_write/file_write -o overwrite -f randomdata.dat -d R```

bunder2015 commented on tests/zfs-tests/tests/functional/rsend/send-c_embedded_blocks.ksh:
```diff
@@ -0,0 +1,103 @@
+	# For lz4, this method works for blocks up to 16k, but not larger
+	[[ $recsize -eq $((32 * 1024)) ]] && break
+
+	log_must mkholes -h 0:$((recsize - 8)) -d $((recsize - 8)):8 \
```
Looks like we're also missing mkholes.

ahrens replied:
`mkholes` is part of the test suite.  see `usr/src/test/zfs-tests/cmd/mkholes` on illumos

bunder2015 replied:
Found it, thanks!

bunder2015 commented on tests/zfs-tests/cmd/mkholes/mkholes.c:
```diff
@@ -0,0 +1,210 @@
+			fl.l_whence = SEEK_SET;
+			fl.l_start = off;
+			fl.l_len = len;
+			if (fcntl(fd, F_FREESP, &fl) != 0) {
```
Will need to find a Linux alternative.

dinatale2 replied:
`fallocate` with the `FALLOC_FL_PUNCH_HOLE` mode might work.

dinatale2 replied:
`tests/zfs-tests/cmd/randfree_file/randfree_file.c` has an example of using `fallocate`.

dinatale2 commented:
@bunder2015 #5952 landed in master. If you rebase, the STYLE builder should begin to pass its style checking.

bunder2015 commented:
Thanks.

Looks like we're down to 3 problems in the test scripts, fio is saying it needs --size, something to do with loop devices on some zpool creates, and a missing include to get within_percent to work.  I'm not 100% on the loop devices but I think I got the other two working.

behlendorf commented:
@bunder2015 my expectation is that #5903 will get merged some time this week.  At which point you won't need to make any of the `zfs` -> `$ZFS` changes in the test suite.

bunder2015 commented:
@dinatale2 might want to check #5952 again, now it says I didn't sign off for a port.  😆

dinatale2 commented:
@bunder2015 It looks like you merged master into your branch. So the buildbot is testing the commit with the message `Merge branch 'master' into illumos-7628`. That's why you are seeing the STYLE builder failing.

What you want to do is roll back to your previous commit (the one that has a commit message of `OpenZFS 7628 - create long versions of ZFS...`). Then you'll want to do a `git rebase` onto master.

bunder2015 commented:
Thanks, I think I got that sorted out.

Looked over the last set of logs and found a few things...
`02:26:04.01 falloc_fl_punch_hole failed: Operation not supported`
Unfortunately, that's not going to work it seems.

`02:26:03.33 NOTE: Comparing 40960 and 28160 given 90% (calculated: 68.75%)`
file_write doesn't seem random enough either.

Missing loop/vdev stuff should be fixed.  Need to look at fio again, doesn't look like its putting its files in the right places.

behlendorf commented:
@bunder2015 hole punching on Linux wasn't supported until the 2.6.38 kernel.  My suggestion would be to update the test case to verify that FALLOC_FL_KEEP_SIZE works and if not skip the test case with `log_unsupported`.  That way we still get good test coverage on the relevant distributions.

loli10K commented:
> falloc_fl_punch_hole failed: Operation not supported

Shouldn't we call

`fallocate(fd, FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE, ...`

instead of

`fallocate(fd, FALLOC_FL_PUNCH_HOLE, ...`

to punch holes? It seems that's what `fallocate(1)` with `--punch-hole` does.

`fallocate`:

```
(gdb) file ./fallocate 
Reading symbols from ./fallocate...done.
(gdb) b fallocate
Breakpoint 1 at 0x401170
(gdb) run --punch-hole -o 256 -l 256 /testpool/fs/urandom 
Starting program: /root/util-linux-2.25.2/fallocate --punch-hole -o 256 -l 256 /testpool/fs/urandom

Breakpoint 1, fallocate (fd=3, mode=3, offset=256, len=140737488349216) at ../sysdeps/unix/sysv/linux/wordsize-64/fallocate.c:28
28	../sysdeps/unix/sysv/linux/wordsize-64/fallocate.c: No such file or directory.
```

`mkholes`:

```
(gdb) file ./bin/mkholes
Reading symbols from ./bin/mkholes...done.
(gdb) b fallocate
Function "fallocate" not defined.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (fallocate) pending.
(gdb) run -h 256:256 /testpool/fs/urandom
Starting program: /usr/share/zfs/zfs-tests/bin/mkholes -h 256:256 /testpool/fs/urandom
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, fallocate (fd=fd@entry=3, mode=mode@entry=2, offset=offset@entry=256, len=140737488349424) at ../sysdeps/unix/sysv/linux/wordsize-64/fallocate.c:28
28	../sysdeps/unix/sysv/linux/wordsize-64/fallocate.c: No such file or directory.
(gdb) p mode
$1 = 2
```

It also seems our `zpl_fallocate_common()` doesn't like `FALLOC_FL_PUNCH_HOLE` without `FALLOC_FL_KEEP_SIZE`, https://github.com/zfsonlinux/zfs/blob/zfs-0.7.0-rc3/module/zfs/zpl_file.c#L659:

```c
   649  zpl_fallocate_common(struct inode *ip, int mode, loff_t offset, loff_t len)
   650  {
   651          int error = -EOPNOTSUPP;
   652  
   653  #if defined(FALLOC_FL_PUNCH_HOLE) && defined(FALLOC_FL_KEEP_SIZE)
   654          cred_t *cr = CRED();
   655          flock64_t bf;
   656          loff_t olen;
   657          fstrans_cookie_t cookie;
   658  
   659          if (mode != (FALLOC_FL_KEEP_SIZE | FALLOC_FL_PUNCH_HOLE))
   660                  return (error);
   661  
```

and the VFS too, http://lxr.free-electrons.com/source/fs/open.c?v=4.6#L244:

```c
244         /* Punch hole must have keep size set */
245         if ((mode & FALLOC_FL_PUNCH_HOLE) &&
246             !(mode & FALLOC_FL_KEEP_SIZE))
247                 return -EOPNOTSUPP;
```

dinatale2 commented:
@loli10K @bunder2015 The man page also confirms the need to have FALLOC_FL_KEEP_SIZE

```
The  FALLOC_FL_PUNCH_HOLE flag must be ORed with FALLOC_FL_KEEP_SIZE in
mode; in other words, even when punching off the end of the  file,  the
file size (as reported by stat(2)) does not change.
```

bunder2015 commented:
I'll take another look, thanks for the tips.  Been busy since Monday, I'll try to get things updated tonight or tomorrow.

bunder2015 commented:
Quick update, rebased on master, fixed fio, fixed file_write, and I think I fixed mkholes, although it got killed by the buildbot last night.  Still need to update embedded_blocks to check for fallocate support.

loli10K commented on tests/zfs-tests/cmd/mkholes/mkholes.c:
```diff
@@ -0,0 +1,255 @@
+
+			buf = (char *)umem_alloc(readlen, UMEM_NOFAIL);
+			vbuf = (char *)umem_zalloc(readlen, UMEM_NOFAIL);
+			while (bytes_read < len) {
```
`fallocate(fd, FALLOC_FL_KEEP_SIZE | FALLOC_FL_PUNCH_HOLE, off, len)` on an empty file won't extend it from `off` of `len` bytes because of `FALLOC_FL_KEEP_SIZE`: `bytes_read` will always be 0 and thus `< len`, this will cause an infinite loop here and is the reason the BB is hitting the 10m limit on `send-c_embedded_blocks`.

loli10K commented on module/zfs/lz4.c:
```diff
@@ -87,7 +87,7 @@ lz4_decompress_zfs(void *s_start, void *d_start, size_t s_len,
 	/*
 	 * Returns 0 on success (decompression function returned non-negative)
-	 * and non-zero on failure (decompression function returned negative.
+	 * and non-zero on failure (decompression function returned negative.)
```
nit: the `.` should be after the parenthesis, not before.

bunder2015 replied:
Oops, thanks.

PS: I got my buildbot running, so I'm hoping to have some of the last few bugs ironed out soon.

loli10K commented on tests/zfs-tests/tests/functional/rsend/rsend.kshlib:
```diff
@@ -551,3 +551,120 @@ function test_fs_setup
+	shift
+
+	[[ -f $file ]] || log_fail "Couldn't find file: $file"
+	typeset flags=$(cat $file | zstreamdump | awk '/features =/ {print $3}')
```
I assume you disabled "send-cpL_varied_recsize" because you're getting "arithmetic syntax error" from ksh.

This seems to happen when we have multiple BEGIN records and the `flags` variable becomes a multiline string:

```
root@debian-8-zfs:~# zfs create $POOLNAME/fs
root@debian-8-zfs:~# zfs create $POOLNAME/fs/sub1
root@debian-8-zfs:~# zfs create $POOLNAME/fs/sub2
root@debian-8-zfs:~# zfs snap -r $POOLNAME/fs@snap1
root@debian-8-zfs:~# zfs send -cpLRDe $POOLNAME/fs@snap1 > $TMPDIR/zstream.dat
root@debian-8-zfs:~# #
root@debian-8-zfs:~# cat $TMPDIR/zstream.dat | zstreamdump | awk '/features =/ {print $3}'
7
430007
430007
430007
root@debian-8-zfs:~# ksh
# flags='7
> 430007
> 430007
> 430007'
# 
# printf "%x" $((0x$flags & 0x430007))
ksh: 0x7
430007
430007
430007 & 0x430007: arithmetic syntax error
# 
```

`mdb` seems to consider only the ending line:

```
[root@52-54-00-d3-7a-f3 ~]# echo -e "0\n0\n4 & 4 = X" | mdb
                4               
[root@52-54-00-d3-7a-f3 ~]# echo -e "0\n4\n0 & 4 = X" | mdb
                0               
[root@52-54-00-d3-7a-f3 ~]# echo -e "4\n0\n0 & 4 = X" | mdb
                0               
[root@52-54-00-d3-7a-f3 ~]# 
```

We should be able to "fix" the issue like this:

```sh
typeset flags=$(cat $file | zstreamdump | awk '/features =/ {features = $3} END {print features}')
```

bunder2015 replied:
It was only intended to be temporary while I ironed out other test issues, but I'll give that a try, thanks.

bunder2015 commented:
Ouch.  send-cpL_varied_recsize gave me this during a zfs receive...

`[202038.186394] BUG: unable to handle kernel paging request at ffffc90003696000
[202038.186513] IP: [<ffffffffa04bc5fe>] dmu_write_impl+0x87/0xc1 [zfs]
[202038.186628] PGD 11f020067 PUD 11f021067 PMD b2268067 PTE 0
[202038.186712] Oops: 0000 [#1] SMP 
[202038.186750] Modules linked in: zfs(PO) icp(PO) zcommon(PO) zunicode(PO) znvpair(PO) zavl(PO) splat(O) spl(O) video backlight [last unloaded: spl]
[202038.186990] CPU: 0 PID: 3689 Comm: receive_writer Tainted: P           O    4.4.52-gentoo #4
[202038.187119] Hardware name: innotek GmbH VirtualBox/VirtualBox, BIOS VirtualBox 12/01/2006
[202038.187243] task: ffff880009db46c0 ti: ffff880019ae8000 task.ti: ffff880019ae8000
[202038.187354] RIP: 0010:[<ffffffffa04bc5fe>]  [<ffffffffa04bc5fe>] dmu_write_impl+0x87/0xc1 [zfs]
[202038.187502] RSP: 0018:ffff880019aebcd8  EFLAGS: 00010286
[202038.187574] RAX: ffff880043400000 RBX: 0000000000020000 RCX: 000000000001f000
[202038.187679] RDX: ffffc9000633f000 RSI: ffffc90003696000 RDI: ffffc90006340000
[202038.187784] RBP: ffff880019aebd28 R08: ffffc90003695000 R09: ffffffffa05ac548
[202038.187889] R10: ffffffff815a1499 R11: 0000000000000001 R12: ffff8800b1e6c730
[202038.187994] R13: 0000000000000003 R14: 00000000000a0000 R15: 0000000000060000
[202038.188025] FS:  0000000000000000(0000) GS:ffff88011fc00000(0000) knlGS:0000000000000000
[202038.188025] CS:  0010 DS: 0000 ES: 0000 CR0: 000000008005003b
[202038.188025] CR2: ffffc90003696000 CR3: 0000000081ec6000 CR4: 00000000000406f0
[202038.188191] Stack:
[202038.188520]  ffffc90003695000 00000008a04bd8fe ffff88010475f380 0000000000000000
[202038.188943]  ffff880070828c40 0000000000100000 0000000000000000 ffff8800a2290000
[202038.189389]  0000000000000007 ffff880019aebd50 ffff880019aebd80 ffffffffa04bda42
[202038.189792] Call Trace:
[202038.190122]  [<ffffffffa04bda42>] dmu_write+0xa9/0xcb [zfs]
[202038.190510]  [<ffffffffa04bdbac>] dmu_assign_arcbuf+0x148/0x166 [zfs]
[202038.190883]  [<ffffffffa04c494b>] receive_write+0x138/0x172 [zfs]
[202038.191245]  [<ffffffffa04c53a0>] ? receive_freeobjects.isra.12+0xa0/0xa0 [zfs]
[202038.191638]  [<ffffffffa04c545f>] receive_writer_thread+0xbf/0x587 [zfs]
[202038.198073]  [<ffffffffa04c53a0>] ? receive_freeobjects.isra.12+0xa0/0xa0 [zfs]
[202038.198441]  [<ffffffffa03d4b5f>] thread_generic_wrapper+0x69/0x73 [spl]
[202038.198795]  [<ffffffffa03d4af6>] ? __thread_exit+0x12/0x12 [spl]
[202038.199142]  [<ffffffff8104bb27>] kthread+0xcd/0xd5
[202038.199471]  [<ffffffff8104ba5a>] ? kthread_create_on_node+0x16d/0x16d
[202038.199824]  [<ffffffff8147c65f>] ret_from_fork+0x3f/0x70
[202038.200154]  [<ffffffff8104ba5a>] ? kthread_create_on_node+0x16d/0x16d
[202038.200518] Code: 49 0f 47 de 48 39 d3 75 07 e8 da b7 ff ff eb 05 e8 20 b5 ff ff 48 8b 55 c8 48 89 d9 49 03 54 24 18 4c 8b 45 b0 48 89 d7 4c 89 c6 <f3> a4 49 3b 5c 24 10 48 89 75 c8 75 0c 48 8b 75 d0 4c 89 e7 e8 
[202038.201748] RIP  [<ffffffffa04bc5fe>] dmu_write_impl+0x87/0xc1 [zfs]
[202038.202108]  RSP <ffff880019aebcd8>
[202038.202407] CR2: ffffc90003696000
[202038.203409] ---[ end trace 62455ade86217e51 ]---`

`[ 3842.517642] BUG: unable to handle kernel paging request at ffffc90008ef9000
[ 3842.517759] IP: [<ffffffffa01175fe>] dmu_write_impl+0x87/0xc1 [zfs]
[ 3842.517870] PGD 11f020067 PUD 11f021067 PMD 96906067 PTE 0
[ 3842.517953] Oops: 0000 [#1] SMP 
[ 3842.517992] Modules linked in: zfs(PO) icp(PO) zcommon(PO) zunicode(PO) znvpair(PO) zavl(PO) splat(O) spl(O) video backlight
[ 3842.518198] CPU: 0 PID: 10727 Comm: receive_writer Tainted: P           O    4.4.52-gentoo #4
[ 3842.518329] Hardware name: innotek GmbH VirtualBox/VirtualBox, BIOS VirtualBox 12/01/2006
[ 3842.518638] task: ffff88005dd551c0 ti: ffff880010f5c000 task.ti: ffff880010f5c000
[ 3842.518823] RIP: 0010:[<ffffffffa01175fe>]  [<ffffffffa01175fe>] dmu_write_impl+0x87/0xc1 [zfs]
[ 3842.519538] RSP: 0018:ffff880010f5fcd8  EFLAGS: 00010286
[ 3842.519909] RAX: ffff8800618e4a00 RBX: 0000000000020000 RCX: 000000000001f000
[ 3842.520368] RDX: ffffc9000b165000 RSI: ffffc90008ef9000 RDI: ffffc9000b166000
[ 3842.520750] RBP: ffff880010f5fd28 R08: ffffc90008ef8000 R09: ffffffffa0207548
[ 3842.521127] R10: ffffffff815a1499 R11: 0000000000000001 R12: ffff880023274e50
[ 3842.521499] R13: 0000000000000003 R14: 00000000000a0000 R15: 0000000000060000
[ 3842.527673] FS:  0000000000000000(0000) GS:ffff88011fc00000(0000) knlGS:0000000000000000
[ 3842.528053] CS:  0010 DS: 0000 ES: 0000 CR0: 000000008005003b
[ 3842.528390] CR2: ffffc90008ef9000 CR3: 00000000b6cba000 CR4: 00000000000406f0
[ 3842.528755] Stack:
[ 3842.529023]  ffffc90008ef8000 00000008a01188fe ffff880117887dc0 0000000000000000
[ 3842.529398]  ffff8800db7547c0 0000000000100000 0000000000000000 ffff88001e88f000
[ 3842.529773]  0000000000000007 ffff880010f5fd50 ffff880010f5fd80 ffffffffa0118a42
[ 3842.530146] Call Trace:
[ 3842.530452]  [<ffffffffa0118a42>] dmu_write+0xa9/0xcb [zfs]
[ 3842.530796]  [<ffffffffa0118bac>] dmu_assign_arcbuf+0x148/0x166 [zfs]
[ 3842.531155]  [<ffffffffa011f94b>] receive_write+0x138/0x172 [zfs]
[ 3842.531507]  [<ffffffffa01203a0>] ? receive_freeobjects.isra.12+0xa0/0xa0 [zfs]
[ 3842.531887]  [<ffffffffa012045f>] receive_writer_thread+0xbf/0x587 [zfs]
[ 3842.532252]  [<ffffffffa01203a0>] ? receive_freeobjects.isra.12+0xa0/0xa0 [zfs]
[ 3842.532617]  [<ffffffffa0014b5f>] thread_generic_wrapper+0x69/0x73 [spl]
[ 3842.532972]  [<ffffffffa0014af6>] ? __thread_exit+0x12/0x12 [spl]
[ 3842.533317]  [<ffffffff8104bb27>] kthread+0xcd/0xd5
[ 3842.533640]  [<ffffffff8104ba5a>] ? kthread_create_on_node+0x16d/0x16d
[ 3842.533998]  [<ffffffff8147c65f>] ret_from_fork+0x3f/0x70
[ 3842.534550]  [<ffffffff8104ba5a>] ? kthread_create_on_node+0x16d/0x16d
[ 3842.534902] Code: 49 0f 47 de 48 39 d3 75 07 e8 da b7 ff ff eb 05 e8 20 b5 ff ff 48 8b 55 c8 48 89 d9 49 03 54 24 18 4c 8b 45 b0 48 89 d7 4c 89 c6 <f3> a4 49 3b 5c 24 10 48 89 75 c8 75 0c 48 8b 
75 d0 4c 89 e7 e8 
[ 3842.536127] RIP  [<ffffffffa01175fe>] dmu_write_impl+0x87/0xc1 [zfs]
[ 3842.536492]  RSP <ffff880010f5fcd8>
[ 3842.536787] CR2: ffffc90008ef9000
[ 3842.537787] ---[ end trace 2a912e8298d08c9e ]---`

loli10K commented:
@bunder2015 that seems to be unrelated (well, at least not caused by) the code in this PR, i can reproduce the same issue with the very first commit that introduced compressed zfs send/receive https://github.com/zfsonlinux/zfs/commit/2aa34383b9362200e19b22ca4a22ea97d70d9ddf. This also happens on Illumos.

More info from `--enable-debug`:
```
ZFS: Loaded module v0.7.0-rc1 (DEBUG mode), ZFS pool version 5000, ZFS filesystem version 5
SPL: using hostid 0x00000000
VERIFY3(arc_get_compression(buf) == ZIO_COMPRESS_OFF) failed (10 == 2)
PANIC at dmu.c:1378:dmu_assign_arcbuf()
Showing stack for process 1927
CPU: 1 PID: 1927 Comm: receive_writer Tainted: P           OE   4.6.4 #5
Hardware name: Bochs Bochs, BIOS Bochs 01/01/2011
 0000000000000286 000000007eeaa7a1 ffffffff8146a545 ffffffffc04668c1
 ffff8800d936bd70 ffffffffc01a19df 0000000000000007 ffffffff00000030
 ffff8800d936bd80 ffff8800d936bd20 2833594649524556 5f7465675f637261
Call Trace:
 [<ffffffff8146a545>] ? dump_stack+0x5c/0x77
 [<ffffffffc01a19df>] ? spl_panic+0xbf/0xf0 [spl]
 [<ffffffff818009b0>] ? __mutex_unlock_slowpath+0xd0/0x180
 [<ffffffff818009b0>] ? __mutex_unlock_slowpath+0xd0/0x180
 [<ffffffffc02cb693>] ? __dbuf_hold_impl+0x173/0xad0 [zfs]
 [<ffffffffc02d9261>] ? dmu_assign_arcbuf+0x121/0x250 [zfs]
 [<ffffffffc02e2c6d>] ? receive_write+0x14d/0x1d0 [zfs]
 [<ffffffffc02e4ae0>] ? receive_writer_thread+0x310/0x840 [zfs]
 [<ffffffff810b03c7>] ? dequeue_task_fair+0x57/0x8e0
 [<ffffffff810b4a7c>] ? pick_next_task_fair+0x10c/0x4a0
 [<ffffffff8102b76f>] ? __switch_to+0x26f/0x530
 [<ffffffff810a1f17>] ? set_user_nice+0x167/0x230
 [<ffffffffc02e47d0>] ? send_traverse_thread+0x140/0x140 [zfs]
 [<ffffffffc019d4ae>] ? thread_generic_wrapper+0x7e/0xc0 [spl]
 [<ffffffffc019d430>] ? __thread_exit+0x20/0x20 [spl]
 [<ffffffff81098ddf>] ? kthread+0xdf/0x100
 [<ffffffff81802eb2>] ? ret_from_fork+0x22/0x40
 [<ffffffff81098d00>] ? kthread_park+0x50/0x50
```

This needs more debugging. Let's keep `send-cpL_varied_recsize` disabled until someone figure this out.

behlendorf commented:
The failures hear are likely related to https://github.com/zfsonlinux/zfs/pull/6003 which @don-brady just opened a PR for.

loli10K commented:
I can still hit the same VERIFY running https://github.com/zfsonlinux/zfs/pull/6003/commits/5800a5b4c2f1999164bf4d044bc5bd539dac8c23:

```
ZFS: Loaded module v0.7.0-rc3_193_g5800a5b (DEBUG mode), ZFS pool version 5000, ZFS filesystem version 5
SPL: using hostid 0x00000000
VERIFY3(arc_get_compression(buf) == ZIO_COMPRESS_OFF) failed (10 == 2)
PANIC at dmu.c:1495:dmu_assign_arcbuf()
Showing stack for process 2133
CPU: 1 PID: 2133 Comm: receive_writer Tainted: P           OE   4.6.4 #5
Hardware name: Bochs Bochs, BIOS Bochs 01/01/2011
 0000000000000286 000000004af464aa ffffffff8146a545 ffffffffc063e3b7
 ffff8800d3667d70 ffffffffc01b59df 0000000000000007 ffffffff00000030
 ffff8800d3667d80 ffff8800d3667d20 2833594649524556 5f7465675f637261
Call Trace:
 [<ffffffff8146a545>] ? dump_stack+0x5c/0x77
 [<ffffffffc01b59df>] ? spl_panic+0xbf/0xf0 [spl]
 [<ffffffff818009b0>] ? __mutex_unlock_slowpath+0xd0/0x180
 [<ffffffffc046c153>] ? __dbuf_hold_impl+0x173/0xad0 [zfs]
 [<ffffffffc0479b61>] ? dmu_assign_arcbuf+0x121/0x250 [zfs]
 [<ffffffffc04846cd>] ? receive_write+0x14d/0x1d0 [zfs]
 [<ffffffffc0486595>] ? receive_writer_thread+0x315/0x850 [zfs]
 [<ffffffff810b03c7>] ? dequeue_task_fair+0x57/0x8e0
 [<ffffffff810b4a7c>] ? pick_next_task_fair+0x10c/0x4a0
 [<ffffffff8102b76f>] ? __switch_to+0x26f/0x530
 [<ffffffff810a1f17>] ? set_user_nice+0x167/0x230
 [<ffffffffc0486280>] ? send_traverse_thread+0x150/0x150 [zfs]
 [<ffffffffc01b14ae>] ? thread_generic_wrapper+0x7e/0xc0 [spl]
 [<ffffffffc01b1430>] ? __thread_exit+0x20/0x20 [spl]
 [<ffffffff81098ddf>] ? kthread+0xdf/0x100
 [<ffffffff81802eb2>] ? ret_from_fork+0x22/0x40
 [<ffffffff81098d00>] ? kthread_park+0x50/0x50
```

Reproducer here: https://gist.github.com/loli10K/5dc583110136d1d4138706555ad1f8a5.

behlendorf commented:
@loli10K @bunder2015 thanks for the reproducer!  The issue is caused by an incorrect check in `do_dump()` regarding when to split large blocks.  It looks like this was caused by a porting mistake in the original patch.  I've added 34c27c9 to this PR which fixes the issue and does some minor updates to the `do_dump()` function to align it with the OpenZFS version.

loli10K commented:
@behlendorf nice, this seems to fix it from my limited testing. I thought this was an issue on the receiving side, the code seems to put a lot of trust (maybe too much) on the input data.

behlendorf commented:
@loli10K yes it does, the recv side could certainly benefit from some additional sanity checks and hardening.

bunder2015 commented:
#6003 and https://github.com/zfsonlinux/zfs/commit/34c27c9a31560ea81ec7ec496aa7227273910651 seem to work for me.  thanks

behlendorf commented:
@bunder2015 feel free to remove the `Requires-builders: none` line from the commit and let the bots do their thing.

bunder2015 commented:
I tried running it through the testbots, but without those 2 patches they all failed.

behlendorf commented:
Replaced by #6067.  @bunder2015 I hope you don't mind but I've taken your work and @don-brady's from #6003 and used them as the basis for #6067.  This updated PR is passing all the new tests after a few minor modifications to the test scripts.  It would be great if you could review it.
